cocos2d/core/renderer/RendererWebGL.js

```diff
@@ -140,16 +259,193 @@ cc.rendererWebGL = {
     },
     
     pushRenderCommand: function (cmd) {
-        if(!cmd._needDraw)
+        if(!cmd.needDraw())
             return;
         if (this._isCacheToBufferOn) {
             var currentId = this._currentID, locCmdBuffer = this._cacheToBufferCmds;
             var cmdList = locCmdBuffer[currentId];
             if (cmdList.indexOf(cmd) === -1)
                 cmdList.push(cmd);
         } else {
-            if (this._renderCmds.indexOf(cmd) === -1)
+            if (this._renderCmds.indexOf(cmd) === -1) {
                 this._renderCmds.push(cmd);
+            }
+        }
+    },
+
+    _increaseBatchingSize: function (increment, vertexType) {
+        vertexType = vertexType || VertexType.QUAD;
+        var i, curr;
+        switch (vertexType) {
+        case VertexType.QUAD:
+            for (i = 0; i < increment; i += 4) {
+                curr = _batchingSize + i;
+                _indexData[_indexSize++] = curr + 0;
+                _indexData[_indexSize++] = curr + 1;
+                _indexData[_indexSize++] = curr + 2;
+                _indexData[_indexSize++] = curr + 1;
+                _indexData[_indexSize++] = curr + 2;
+                _indexData[_indexSize++] = curr + 3;
+            }
+            break;
+        case VertexType.TRIANGLE:
+            _pureQuad = false;
+            for (i = 0; i < increment; i += 3) {
+                curr = _batchingSize + i;
+                _indexData[_indexSize++] = curr + 0;
+                _indexData[_indexSize++] = curr + 1;
+                _indexData[_indexSize++] = curr + 2;
+            }
+            break;
+        default:
+            return;
+        }
+        _batchingSize += increment;
+    },
+
+    _uploadBufferData: function (cmd) {
+        if (_batchingSize >= _maxVertexSize) {
+            this._batchRendering();
+        }
+
+        // Check batching
+        var texture = cmd._texture || cmd._node._texture;
+        var blendSrc = cmd._node._blendFunc.src;
+        var blendDst = cmd._node._blendFunc.dst;
+        var shader = cmd._shaderProgram;
```

Comment:
Need to find a better way for finding those properties, at lease _texture is not very reliable.